v2/pkg/astnormalization/uploads/upload_finder.go

```diff
@@ -0,0 +1,270 @@
+	v.currentArgPath.reset()
+	v.currentVariablePath.reset()
+	v.variableUsedDirectlyOnArg = false
+	v.uploadPathMapping = v.uploadPathMapping[:0]
+}
+
+func (v *UploadFinder) FindUploads(operation, definition *ast.Document, variables []byte, argRef int, argInputValueDefinitionRef int) (uploadPathMapping []UploadPathMapping, err error) {
+	v.definition = definition
+	v.operation = operation
+
+	if variables == nil || bytes.Equal(variables, []byte("null")) || bytes.Equal(variables, []byte("")) {
+		variables = []byte("{}")
+	}
+
+	v.variables, err = astjson.ParseBytesWithoutCache(variables)
+	if err != nil {
+		return nil, err
+	}
+
+	v.currentArgPath.reset()
+	v.currentVariablePath.reset()
+
+	argumentValue := v.operation.Arguments[argRef].Value
+	argumentTypeRef := v.definition.InputValueDefinitionType(argInputValueDefinitionRef)
+
+	v.traverseValue(argumentValue, argumentTypeRef)
+
+	if len(v.uploadPathMapping) > 0 {
+		return v.uploadPathMapping, nil
+	}
+
+	return nil, nil
+}
+
+func (v *UploadFinder) traverseValue(value ast.Value, valueTypeRef int) {
+	switch value.Kind {
+	case ast.ValueKindVariable:
+		v.variableUsedDirectlyOnArg = !v.currentArgPath.hasPath()
+		v.traverseVariable(value.Ref, valueTypeRef)
+	case ast.ValueKindList:
+		listItemTypeRef := valueTypeRef
+		for i, ref := range v.operation.ListValues[value.Ref].Refs {
+			v.currentArgPath.pushArrayPath(i)
+			v.traverseValue(v.operation.Value(ref), v.definition.Types[listItemTypeRef].OfType)
+			v.currentArgPath.popPath()
+		}
+	case ast.ValueKindObject:
+		typeNameBytes := v.definition.ResolveTypeNameBytes(valueTypeRef)
+		objectTypeNode, ok := v.definition.NodeByName(typeNameBytes)
+		if !ok {
+			return
+		}
+		if objectTypeNode.Kind != ast.NodeKindInputObjectTypeDefinition {
+			return
+		}
+
+		for _, ref := range v.operation.ObjectValues[value.Ref].Refs {
+			objectFieldName := v.operation.ObjectFieldNameBytes(ref)
+
+			fieldInputValueDefinitionRef, ok := v.definition.InputValueDefinitionRefByInputObjectDefinitionRefAndFieldNameBytes(objectTypeNode.Ref, objectFieldName)
+			if !ok {
+				continue
+			}
+
+			v.currentArgPath.pushObjectPath(objectFieldName)
+			v.traverseValue(v.operation.ObjectField(ref).Value, v.definition.InputValueDefinitionType(fieldInputValueDefinitionRef))
+			v.currentArgPath.popPath()
+		}
+	case ast.ValueKindNull:
+		// we care only about null - because it could be an upload
+		if v.definition.ResolveTypeNameString(valueTypeRef) == "Upload" {
+			v.AddUploadPath()
+		}
+	}
+}
+
+func (v *UploadFinder) traverseVariable(variableValueRef int, definitionValueTypeRef int) {
+	operationVariableNameBytes := v.operation.VariableValueNameBytes(variableValueRef)
+	operationVariableName := unsafebytes.BytesToString(operationVariableNameBytes)
+
+	v.currentVariableName = operationVariableNameBytes
+	v.currentVariableValue = v.variables.Get(operationVariableName)
+
+	// we rely on the fact that there is only one operation in the document
```

Comment:
Is this the right approach?
From my understanding, e.g. persisted operations can have multiple ops per doc

Reply:
after normalization we should always have a single operation